packages/pages-shared/asset-server/handler.ts

```diff
@@ -199,9 +199,9 @@ export async function generateHandler<
 					? `${destination.pathname}${destination.search || search}${
 							destination.hash
 					  }`
-					: `${destination.href}${destination.search ? "" : search}${
-							destination.hash
-					  }`;
+					: `${destination.origin}${destination.pathname}${
```

Comment:
any reason to not just use `destination.href` and leave it at that?

Reply:
Agreed - if we want the entire URL, we can use `.href` or `.toString()` (the latter is a bit clearer in my mind, but both do the same thing.)
[Playground example](https://www.typescriptlang.org/play?ssl=6&ssc=70&pln=6&pc=18#code/MYewdgzgLgBArgJwDYwLwzAUwO4wKoBKAMgBQBEAFlFAA4QBcA9I5gB4CGAtjUpgHShOjAGYgQjAEbsEAfmEI4ASyip2NHpgBkAEwSKwAa1QArJcEwBiCXACeZAJQAoR6EghefJCADmJREj4KBExhJ1cId34vX38+KBAAZSg9MF97J0dmGAAhECgKGBoU2CpaBmY2Lg0BECFRcSlZeSUVNQ0dFKNTRXMrWyA)

Reply:
Agreed - if we want the entire URL, we can use `.href` or `.toString()` (the latter is a bit clearer in my mind, but both do the same thing.)
[Playground example](https://www.typescriptlang.org/play?ssl=6&ssc=70&pln=6&pc=18#code/MYewdgzgLgBArgJwDYwLwzAUwO4wKoBKAMgBQBEAFlFAA4QBcA9I5gB4CGAtjUpgHShOjAGYgQjAEbsEAfmEI4ASyip2NHpgBkAEwSKwAa1QArJcEwBiCXACeZAJQAoR6EghefJCADmJREj4KBExhJ1cId34vX38+KBAAZSg9MF97J0dmGAAhECgKGBoU2CpaBmY2Lg0BECFRcSlZeSUVNQ0dFKNTRXMrWyA)

Reply:
Because sometimes we want to replace the `URL.search` component (and any `URL.hash` component follows the `URL.search` component).

Granted, we could manipulate `destination` itself e.g. `destination.search = destination.search ? "" : search`, but that's not _great_ either.

Reply:
I've always modified parts of the url and then done `toString()` it at the end, which feels more straightforward and easier to understand the intent. The way it's written now, it took me a bit to understand what we were trying to do.
I'd go one step further to make this clearer:
```
if(!destination.search) {
    destination.search = search
}
... = destination.toString()
```

Reply:
I agree with Jacob. Re-implementing URL stringification was what led to this bug in the first place and will likely save us in the future (barring any wonky `URL` implementations in the runtime).

My rationale for using Greg's suggestion was that it was the minimum viable change for the bug fix

Reply:
I'm fine with doing that if @jrf0110 prefers it too.

I generally shy away from manipulating parts of objects like this, just because the meaning of the variable can get a bit fuzzy.

In a very contrived example, here, because we manipulate `shoppingCart`, it's state is somewhat unclear at each point.

```ts
const shoppingCart = getExistingShoppingCart(cookie)
// { total: 0, items: [] }

shoppingCart.items = [...shoppingCart.items, { title: "Book", price: 12 }]
shoppingCart.total = shoppingCart.total + 12

saveShoppingCart(shoppingCart, cookie)
```

Versus here, where we create new variables whenever it gets new meaning:

```
const existingShoppingCart = getExistingShoppingCart(request)
// { total: 0, items: [] }

const updatedShoppingCart = {
  items: [...existingShoppingCart, { title: "Book", price: 12 }],
  total: existingShoppingCart + 12,
}

saveShoppingCart(shoppingCart, cookie)
```

If I were to inspect the state of the world at any point in this flow, the variables would always have good descriptions whereas in the first one, I can't immediately tell the state of the shopping cart. More wasteful with memory, but less bug-prone in my experience.

But back to this particular situation... this is really just a minor signposting of intention really, and it's a well-tested codebase, so I'm honestly happy with either way.

Reply:
Thanks for the example @GregBrimble - that's a great point. I think there's definitely a balance we can strike here. I think as long as we're only modifying the URL in a self-contained manor it feels pretty safe. But we should be careful that we aren't modifying a given url object across a lot of locations.

One thing we could do to help mitigate this risk is split this functionality out into a function so that the modifications are fully self-contained.

Something like `location = transformRedirectDestination(destination)`